server/datastore/mysql/apple_mdm.go

```diff
@@ -231,3 +235,304 @@ WHERE
 	}
 	return devices, nil
 }
+
+func (ds *Datastore) IngestMDMAppleDeviceFromCheckin(ctx context.Context, mdmHost fleet.MDMAppleHostDetails) error {
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		return ingestMDMAppleDeviceFromCheckinDB(ctx, tx, mdmHost)
+	})
+}
+
+func ingestMDMAppleDeviceFromCheckinDB(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	mdmHost fleet.MDMAppleHostDetails,
+) error {
+	if mdmHost.SerialNumber == "" {
+		return ctxerr.New(ctx, "ingest mdm apple host from checkin expected device serial number but got empty string")
+	}
+	if mdmHost.UDID == "" {
+		return ctxerr.New(ctx, "ingest mdm apple host from checkin expected unique device id but got empty string")
+	}
+
+	stmt := `SELECT id, uuid, hardware_serial FROM hosts WHERE uuid = ? OR hardware_serial = ?`
+
+	var foundHost fleet.Host
+	err := sqlx.GetContext(ctx, tx, &foundHost, stmt, mdmHost.UDID, mdmHost.SerialNumber)
+	switch {
+	case err != nil && !errors.Is(err, sql.ErrNoRows):
+		return ctxerr.Wrap(ctx, err, "get mdm apple host by serial number or udid")
+
+	case errors.Is(err, sql.ErrNoRows):
+		return insertMDMAppleHostDB(ctx, tx, mdmHost)
+
+	case foundHost.HardwareSerial != mdmHost.SerialNumber || foundHost.UUID != mdmHost.UDID:
+		return updateMDMAppleHostDB(ctx, tx, foundHost.ID, mdmHost)
+
+	default:
+		// ok, nothing to do here
+		return nil
+	}
+}
+
+func updateMDMAppleHostDB(ctx context.Context, tx sqlx.ExtContext, hostID uint, mdmHost fleet.MDMAppleHostDetails) error {
+	updateStmt := `
+		UPDATE hosts SET
+			hardware_serial = ?,
+			uuid = ?,
+			hardware_model = ?,
+			platform =  ?,
+			refetch_requested = ?,
+			osquery_host_id = ?
+		WHERE id = ?`
+
+	if _, err := tx.ExecContext(
+		ctx,
+		updateStmt,
+		mdmHost.SerialNumber,
+		mdmHost.UDID,
+		mdmHost.Model,
+		"darwin",
+		1,
+		mdmHost.UDID,
+		hostID,
+	); err != nil {
+		return ctxerr.Wrap(ctx, err, "update mdm apple host")
+	}
+
+	return nil
+}
+
+func insertMDMAppleHostDB(ctx context.Context, tx sqlx.ExtContext, mdmHost fleet.MDMAppleHostDetails) error {
+	insertStmt := `
+		INSERT INTO hosts (
+			hardware_serial,
+			uuid, 
+			hardware_model, 
+			platform, 
+			last_enrolled_at, 
+			detail_updated_at, 
+			osquery_host_id,
+			refetch_requested
```

Comment:
You're right the way these timestamps are being handled in different scenarios gets quite confusing. I think we need more clarity from the interface team regarding how they will be using `last_enrolled_at`, `detail_updated_at` and other timestamp values. My working understanding is that the intent is to use the `last_enrolled_at` value to filter pending MDM hosts.

Reply:
can we set the timestamp to `1970-01-02 01:00:00` to match what Orbit does? or are we using different timestamps to differentiate between the two mechanisms to enroll?

while this is not ideal it can be refactored as a whole later on.

Reply:
I was matching the default values for `label_updated_at` and `policy_updated_at` as those seemed closer to this use case where the UI would be displaying a pending state. I don't know why Orbit is using the date that it is. In my opinion, moving in the direction of consistency with the label and policy timestamps makes the most sense.

Reply:
that makes sense to me, worst case scenario we can come back to this, let's just add a comment in both methods.